Assert edit permissions on appropriate mutations in namespaceUtils, fix promise logic when deploying a model or adding a model server, show inner error messages from the backend when present

mturley opened the pull request:
Resolves [RHOAIENG-548](https://issues.redhat.com/browse/RHOAIENG-548) and [RHOAIENG-556](https://issues.redhat.com/browse/RHOAIENG-556).

<!--- If this is a non-code change, this template is not required; reference any issues or top-level descriptions as needed -->
<!--- All code change PRs should relate to an issue, reference it here; see example below -->
<!--- Closes: #123 -->

## Description
<!--- Describe your changes in detail; the what, the why, any findings, etc -->
<!--- Include any screenshots of changed UI; Include any gifs if it was a flow / UX change -->

When a user has "edit" permission on a data science project they don't own, they don't have permission to update that namespace. When adding the first model or model server to a project, the namespace needs to be updated by `addSupportServingPlatformProject` to add labels that describe which serving platform the project is using, which fails with a 403 error for users in this position (as expected). Currently, even though this namespace update fails, the ServingRuntime gets created anyway. It does not become visible in the UI because it still shows the platform-selection view (choosing single- or multi-model serving), but it is visible in the cluster.

The `submitServingRuntimeResources` function that is called to make these updates and create the ServingRuntime currently operates like this:
* Execute a dry-run of the operations concurrently with a `Promise.all`.
* If no error in the dry-run, execute the same operations with another `Promise.all`, but this time include the namespace update (because the namespace update performed by `addSupportServingPlatformProject` has no dry-run option).

This means that if `addSupportServingPlatformProject` fails, the other requests will run anyway. The fix is to add a step in between the dry-run and the actual run of these requests where we perform the namespace update before proceeding, which means if it fails the ServingRuntime will not be created. As part of the fix, we convert the `submitServingRuntimeResources` to an async function so we can use `await` for readability.

There are also a few other notable changes:
* Removes the `allowCreate` condition on updating the namespace, which was a relic of an earlier workaround that is no longer needed (@lucferbux can provide more context on this -- thanks for your help Lucas!).
* Changes the `applyNamespaceChange` function on the backend to perform a more accurate `selfSubjectAccessReview`: users with "edit" permission should still be able to create ServingRuntimes when the namespace update is not needed (a serving platform has already been chosen for the project), so we check for that permission only for the `MODEL_MESH_PROMOTION` and `KSERVE_PROMOTION` cases.
* When deploying a Kserve model, there was another `Promise.all` for concurrently calling `submitServingRuntimeResources` and `submitInferenceServiceResource`, which means if the error described above happens here the InferenceService will also still be created. Changing this to be a sequential `.then` also fixes that problem. (Edit: I hadn't realized this was the separate issue [RHOAIENG-556](https://issues.redhat.com/browse/RHOAIENG-556), so this PR fixes that now too :))
* Adds a new `throwErrorFromAxios` utility and uses it in `addSupportServingPlatformProject` (and in `createProject`, the only other place we're directly calling `axios()`). This will ensure that if an error thrown by the `axios()` call has a response object containing a message from the backend, that message will be used instead of the top-level `message` property from Axios ("Request failed with status code X"). If the error being displayed does not contain an inner error message in this expected structure, the current behavior is retained and the top-level `error.message` is used.

Considering that last point above, the error message that used to look like this:

![Screenshot 2024-01-15 at 5 44 47 PM](https://github.com/opendatahub-io/odh-dashboard/assets/811963/29235ed7-ead9-408b-abc5-b1ad3ef4ebb5)

now looks like this:

![Screenshot 2024-01-16 at 3 07 39 PM](https://github.com/opendatahub-io/odh-dashboard/assets/811963/8f6e23e3-d7c1-4752-b63a-7632497efb9b)

@vconzola, does that error text look good to you?

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

With @lucferbux's help, I reproduced the error in a cluster and verified the fix by using the PR image in that cluster. Testing locally gave us trouble because apparently an impersonated regular user cannot list templates in a namespace they don't own (which does not seem to match the behavior in production), so we can't get the "Models and model servers" section of the data science project page to render when trying to reproduce the bug this way.

## Test Impact
<!--- What tests have you done to covert implemented functionality -->
<!--- If tests are not applicable, explain why here -->

New Cypress tests have been added to submit the "deploy model" and "add model server" modals and intercept their requests, ensuring that the request to create ServingRuntimes and InferenceServices are not made when there is an error updating the namespace.

## Request review criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit tests & storybook for related changes)

If you have UI changes: 
<!--- You can ignore these if you are doing manifest, backend, internal logic, etc changes; aka non-UI / visual changes -->
- [x] Included any necessary screenshots or gifs if it was a UI change.
- [x] Included tags to the UX team if it was a UI/UX change (find relevant UX in the [SMEs](https://github.com/opendatahub-io/odh-dashboard/tree/main/docs/smes.md) section).

After the PR is posted & before it merges:
- [x] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

lucferbux commented on backend/src/routes/api/namespaces/namespaceUtils.ts:
```diff
@@ -32,6 +27,34 @@ const checkNamespacePermission = (
   });
 };
 
+const checkNamespacePermission = (
```
Just a nit, you can call this `checkAdminNamespacePermission` or something if you want.

lucferbux commented on frontend/package-lock.json:
```diff
@@ -22,7 +22,6 @@
         "@patternfly/react-table": "^5.1.1",
         "@patternfly/react-topology": "^5.1.0",
         "@patternfly/react-virtualized-extension": "^5.0.0",
-        "@storybook/test-runner": "^0.16.0",
```
Is this intentional? Pretty weird if we have change something in the package json that the package-lock didn't get it.

andrewballantyne replied:
I don't think you should ever modify the package-lock like this... What's the solution being applied here?

mturley replied:
This wasn't intentional, thanks for catching it. I just ran `npm install` and it snuck in for some reason. It looks like rebasing and re-running install has undone it.

mturley replied:
This is what I get for using `git commit -a` 🙃

lucferbux reviewed (changes requested):
Everything looks good to me, just take a look at the package-lock.json situation (I mean, it could be something expected but it's kinda weird).
You can undraft it if you want!

mturley commented:
Sorry for the delay here, I'm back from PTO. @lucferbux I only still have this in draft because I figured it should have updated unit tests before it's ready to merge. Getting back to that now.

mturley commented:
Hey @andrewballantyne @lucferbux, quick question here: I've been poking around and reading code trying to figure out how to mock things correctly to test the `submitServingRuntimeResources` function in `frontend/src/pages/modelServing/screens/projects/__tests__/utils.spec.ts` by somehow mocking a user with and without the appropriate permissions and then calling the function and inspecting the returned promise(s). Do you know if I'm on the right track there? I'm a bit stuck on that, and it may be telling that there aren't any tests for that function yet. Maybe instead I should add an integration test for `ManageServingRuntimeModal` or `ManageKServeModal`. Or I wonder if we should just let this PR merge without new tests.

andrewballantyne commented:
We don't have e2e tests... why do you need to mock user permissions? @christianvogt can you help out here?

mturley commented:
It also looks like the accessibility tests are failing to run in CI but they pass locally :(

mturley commented:
> We don't have e2e tests... why do you need to mock user permissions? @christianvogt can you help out here?

I figured since the fix here has to do with behavior being different for users with different permissions, I would want to mock each of the two cases (should create model server if user can update the namespace, should not create model server if user cannot update the namespace). My apologies for my inexperience with the repo getting in the way here.

christianvogt commented:
@mturley as far as mocking goes. Use jest to mock whatever you need on the frontend being used by the function you're testing. It's sometimes best to mock down to the k8s apis from `@openshift/dynamic-plugin-sdk-utils` - eg `k8sListResource`, `k8sGetResource`, etc...

But sometimes those mocks end up being very deep in the call tree and therefore you may want to mock another function that's used directly by your function. The choice is yours.

The other option is to add a new cypress test where you'd end up mocking the network calls.

mturley commented:
/test all

lucferbux commented:
@mturley @christianvogt I wouldn't wast too much time trying to mock e2e within the backend, the main point i would cover here would be [this call](https://github.com/opendatahub-io/odh-dashboard/pull/2319/files#diff-6f3c32a5088b5dd741e08b771cacb6c4187fcb2d5168d3c4d0caca154bda635eR398-R413). Just try to mock it making sure it's failing so we can prove the rest of the requests are not being executed.

mturley commented:
@lucferbux I believe this is now all set with everything we discussed. I wrote up as much detail as I could in the PR description.

For now, I kept the error message improvement as a separate commit because it does impact all the modals that use `DashboardModalFooter` to display errors. I'm curious if any of you have thoughts on that, or whether it should be a separate PR. cc @andrewballantyne @christianvogt

Also, @vconzola I need to run the new error message text by you, see screenshots in the description above. Does that look good?

lucferbux commented on frontend/src/concepts/dashboard/DashboardModalFooter.tsx:
```diff
@@ -32,7 +43,7 @@ const DashboardModalFooter: React.FC<DashboardModalFooterProps> = ({
       <StackItem>
         <Alert isInline variant="danger" title={alertTitle}>
-          {error.message}
+          {getErrorMessage(error)}
```
Mmmmm great addition, I would even say that the `isAxiosErrorWithInnerMessage` could be a util method in other components, there are several places where we use axios, not sure if we are handling well this, we can further investigate and log something if that's the case.

mturley replied:
Yeah, just searching the repo for `error.message` reveals a number of places where this might be useful. I wonder what the best way is to follow up on that after this PR, do you think it warrants a Jira issue?

lucferbux commented on frontend/src/pages/modelServing/screens/projects/ServingRuntimeModal/ManageServingRuntimeModal.tsx:
```diff
@@ -1,25 +1,22 @@
+import { AxiosError } from 'axios';
 import {
   Alert,
-  Button,
```
Ok, I'll add the comment here, I think I explained myself here wrong yesterday, I do apologize, we still need the accessReviewResource and all of his logic, the only thing we needed to do was remove it from the serving runtime conditional

mturley replied:
Ok, sorry for the misunderstanding. I thought the issue was that the `useAccessReview` itself wouldn't work properly for the restricted user.

lucferbux commented on frontend/src/pages/modelServing/screens/projects/kServeModal/ManageKServeModal.tsx:
```diff
@@ -145,28 +146,27 @@ const ManageKServeModal: React.FC<ManageKServeModalProps> = ({
-        acceleratorProfileState,
-      ),
-    ])
+    submitServingRuntimeResources(
```
Yes, that's the solution I was thinking about for https://issues.redhat.com/browse/RHOAIENG-556, you've solved it in this PR, so attatch the link in the pr description and update the jira issue accordingly, great work here!

mturley replied:
Done!

lucferbux commented on frontend/src/pages/modelServing/screens/projects/utils.ts:
```diff
@@ -340,26 +339,19 @@ export const submitServingRuntimeResources = (
     ...(name !== undefined && { name }),
   };
   const servingRuntimeName = translateDisplayNameForK8s(servingRuntimeData.name);
-  const createRolebinding = servingRuntimeData.tokenAuth && allowCreate;
```
Yeah we still need this!

lucferbux reviewed (changes requested):
Ok, I do apologize for the misunderstanding regarding the allow create, revert that part back and it should be ready to go 💪

mturley commented:
@lucferbux sorry for the misunderstanding there. Reverted those changes, should be all set now I think.

andrewballantyne commented on backend/src/routes/api/namespaces/namespaceUtils.ts:
```diff
@@ -47,40 +70,48 @@ export const applyNamespaceChange = async (
+        checkPermissionsFn = checkEditNamespacePermission;
+      }
       break;
     default:
       throw createCustomError('Unknown configuration', 'Cannot apply namespace change', 400);
   }
 
+  const selfSubjectAccessReview = await checkPermissionsFn(fastify, request, name);
+  if (isK8sStatus(selfSubjectAccessReview)) {
```
It'll fall into the next if-statement after this -- but I am wondering if we want to put a "dev setup" if statement first -- that basically says "`checkPermissionsFn` is null -- you need to set permissions for all actions" -- basically as a "fail immediately" sorta deal with a dev message so the dev knows the new case they added needs to have a check permissions.

Just my two cents. It should fail in some fashion on the `checkPermissionsFn()` call but I dunno how that will look.

mturley replied:
Ah, fair enough. Just another `throw createCustomError` maybe? It would be impossible to hit with the current code but that's a good point about future switch cases. Alternatively though we could just have it fall back to `checkAdminNamespacePermission` which is the behavior in all cases before this PR? I'm leaning towards your idea.

andrewballantyne replied:
Yeah, like a 500 error "invalid backend state -- dev broken workflow" or something like that. Something annoying and blunt it was a dev setup. It will be easy to search in the codebase and should prevent anything from being poorly configured. 

It should never happen in production, but it won't mask when code is poorly written -- so it gets caught in the PR for the change. Or in our test infra when we have that going nightly or whatever.

mturley replied:
Pushed with this change as well.

andrewballantyne commented on frontend/src/concepts/dashboard/DashboardModalFooter.tsx:
```diff
@@ -15,9 +16,19 @@ type DashboardModalFooterProps = {
+  error?: Error | AxiosError;
 };
 
+const isAxiosErrorWithInnerMessage = (
```
Can we turn this into a parse layer at the axios usage -- I'd like to drop axios and I don't think the world of components needs to know about axios' internal usage. Just throw a new error in the `.catch` with a parsed value.

andrewballantyne replied:
I'd imagine you could even make a utility `throwErrorFromAxios` which parses the error out of the `.catch` and then throws it from the utility -- which should still bubble back up.

Something like:
```
axios(...)
  .then(() => /* some then functionality */)
  .catch(throwErrorFromAxios)
```

mturley replied:
Hmm, yeah this does seem like a better separation of concerns. It won't have the benefit of improving the other errors outside the scope of these namespace updates, but maybe that's a good thing. I'm also hesitant to make that change for all axios usages, at least in this PR. It might be worth trying that in a standalone PR where we can easily revert it if it causes trouble elsewhere.

andrewballantyne replied:
That's fair -- don't worry about the wide axios usage -- let us just go with this particular call and log a Task issue for tech debt cleanup.

Ideally I'd like to drop axios in the long run, as it predates my chosen path and eventually I'd like to lean 100% into proxy utils and the SDK utils... prometheus calls also are narrow now that we have UWM stuff setup from the stuff Alex did. Axios to call our node instance for custom business logic should go away in https://issues.redhat.com/browse/RHOAIENG-1461

mturley replied:
@andrewballantyne I think maybe for this `throwErrorFromAxios` utility we may not want to entirely discard the outer AxiosError object and throw only the inner error response, because I'm not sure what other code might be expecting a true `AxiosError` with all its properties. What about instead having the utility just look for an inner `message` if present and rewrite the `message` at the top, retaining the rest of the `AxiosError`? I'll play with that and push a change to show you what I mean.

mturley replied:
Pushed. I added the following to `frontend/src/api/errorUtils.ts`: 

```ts
const isAxiosErrorWithResponseMessage = (
  error?: Error | AxiosError,
): error is AxiosError<{ message: string }> =>
  Boolean(
    error && typeof (error as AxiosError<{ message: string }>).response?.data?.message === 'string',
  );

export const throwErrorFromAxios = (error?: Error | AxiosError) => {
  if (isAxiosErrorWithResponseMessage(error)) {
    throw { ...error, message: error.response?.data.message };
  }
  throw error;
};
```

and then used it in the `catch` as you described above, but only for this one request.

mturley replied:
Hmm, it looks like we only actually call `axios()` directly in one other place: when creating a data science project. I may as well use this there too.

andrewballantyne replied:
Could we throw `new Error` like objects? And try to not throw the Axios error unless it has a `.message` on it

mturley replied:
I wasn't sure if we wanted to clobber anything else on the existing error object that we might want in the components, but I suppose we're only really reading the `.message` anyway. I'll change it so it throws a `new Error` with the response data object only, if present, and otherwise falls back to throwing the AxiosError

andrewballantyne commented on frontend/src/pages/modelServing/screens/projects/ServingRuntimeModal/ManageServingRuntimeModal.tsx:
```diff
@@ -65,7 +66,7 @@ const ManageServingRuntimeModal: React.FC<ManageServingRuntimeModalProps> = ({
     useServingAcceleratorProfile(editInfo?.servingRuntime);
   const [actionInProgress, setActionInProgress] = React.useState(false);
-  const [error, setError] = React.useState<Error | undefined>();
+  const [error, setError] = React.useState<Error | AxiosError | undefined>();
```
My last comment will make this obsolete.

vconzola commented:
@mturley Can you make the error message more specific by replacing "labels" with "serving platform labels". A normal DS might not know what a label is, but they'll know it has something to do with a serving platform and their permissions. Also, put a period at the end.

mturley commented:
@andrewballantyne I've rebase and pushed with changes to address your feedback. Updated the PR description to match. Let me know what you think of the new error message approach.

@vconzola you got it, updated the message and the screenshot.

mturley commented:
Fixed some lint errors I missed. Should be ready for re-review.

mturley commented:
Rebased and addressed remaining comments.

lucferbux commented on backend/src/routes/api/namespaces/namespaceUtils.ts:
```diff
@@ -47,40 +70,55 @@ export const applyNamespaceChange = async (
 
+  if (checkPermissionsFn === null) {
+    throw createCustomError(
+      'Invalid backend state -- dev broken workflow',
```
Yes, great addition here, good catch

lucferbux reviewed (approved):
/lgtm

Ok, after andrew's suggestion, I think is ready to go, the message is descriptive enough to get a grasp of what's happening. Works great in both kserve and modelmesh, great work!

<img width="815" alt="Screenshot 2024-01-17 at 18 27 11" src="https://github.com/opendatahub-io/odh-dashboard/assets/16117276/39c7af3f-4687-4473-9cc2-4429544d0327">
